File: testAPI.py
#!/usr/bin/env python3
"""
Push test sensor readings to the API for visualization.

Behavior:
- Total duration: ~10 seconds
- Number of pushes: 3
- Interval: 3.3333 seconds

Usage:
    python testAPI.py
    python testAPI.py --base-url http://localhost:3000 --location-id 4
"""

from __future__ import annotations

import argparse
import json
import time
import urllib.error
import urllib.request
from typing import Any


TOTAL_SECONDS = 10.0
PUSH_COUNT = 3
INTERVAL_SECONDS = 3.3333


def post_json(url: str, payload: dict[str, Any], timeout: int = 10) -> tuple[int, str]:
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url=url,
        data=data,
        method="POST",
        headers={"Content-Type": "application/json"},
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            body = response.read().decode("utf-8")
            return response.status, body
    except urllib.error.HTTPError as err:
        return err.code, err.read().decode("utf-8", errors="replace")
    except Exception as err:
        return 0, str(err)


def main() -> int:
    parser = argparse.ArgumentParser(description="Push test readings to /api/sensors")
    parser.add_argument("--base-url", default="http://localhost:3000")
    parser.add_argument("--location-id", default="4")
    args = parser.parse_args()

    endpoint = f"{args.base_url.rstrip('/')}/api/sensors"
    start = time.time()

    print(f"Sending test data to: {endpoint}")
    print(f"Schedule: {PUSH_COUNT} pushes in ~{TOTAL_SECONDS}s (every {INTERVAL_SECONDS}s)\n")

    for i in range(1, PUSH_COUNT + 1):
        target_elapsed = INTERVAL_SECONDS * i
        now_elapsed = time.time() - start
        sleep_for = max(0.0, target_elapsed - now_elapsed)
        time.sleep(sleep_for)

        # Change values each push so graph/cards visibly update.
        ultrasonic_value = 15.0 + i * 4.5
        rain_value = 10.0 + i * 8.0
        temp_value = 28.0 + i * 0.9

        packets = [
            {"location_id": args.location_id, "type": "Ultrasonic", "value": ultrasonic_value},
            {"location_id": args.location_id, "type": "Rain Intensity", "value": rain_value},
            {"location_id": args.location_id, "type": "Temperature", "value": temp_value},
        ]

        print(f"Push {i}/{PUSH_COUNT} at t={time.time() - start:.2f}s")
        for packet in packets:
            status, body = post_json(endpoint, packet)
            ok = 200 <= status < 300
            marker = "OK" if ok else "ERR"
            print(f"  {marker} {status} -> {packet}")
            print(f"    response: {body}")
        print()

    total = time.time() - start
    print(f"Finished in {total:.2f}s")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: testAPI.ts
/**
 * Push test sensor readings to the API for visualization.
 *
 * Behavior:
 * - Total duration: ~10 seconds
 * - Number of pushes: 3
 * - Interval: 3.3333 seconds
 *
 * Usage:
 *   npx ts-node testAPI.ts
 *   npx ts-node testAPI.ts --base-url=http://localhost:3000 --location-id=4
 */

type Args = {
  baseUrl: string;
  locationId: string;
};

const TOTAL_SECONDS = 10.0;
const PUSH_COUNT = 3;
const INTERVAL_SECONDS = 3.3333;

function parseArgs(): Args {
  const args = process.argv.slice(2);
  const out: Args = {
    baseUrl: 'http://localhost:3000',
    locationId: '4',
  };

  for (const arg of args) {
    if (arg.startsWith('--base-url=')) out.baseUrl = arg.replace('--base-url=', '');
    if (arg.startsWith('--location-id=')) out.locationId = arg.replace('--location-id=', '');
  }

  return out;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function postJson(url: string, payload: unknown): Promise<{ status: number; body: string }> {
  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    const body = await res.text();
    return { status: res.status, body };
  } catch (error) {
    return { status: 0, body: String(error) };
  }
}

async function main() {
  const { baseUrl, locationId } = parseArgs();
  const endpoint = `${baseUrl.replace(/\/$/, '')}/api/sensors`;
  const start = Date.now();

  console.log(`Sending test data to: ${endpoint}`);
  console.log(`Schedule: ${PUSH_COUNT} pushes in ~${TOTAL_SECONDS}s (every ${INTERVAL_SECONDS}s)\n`);

  for (let i = 1; i <= PUSH_COUNT; i++) {
    const targetElapsedMs = INTERVAL_SECONDS * i * 1000;
    const nowElapsedMs = Date.now() - start;
    const waitMs = Math.max(0, targetElapsedMs - nowElapsedMs);
    await sleep(waitMs);

    const ultrasonicValue = 15.0 + i * 4.5;
    const rainValue = 10.0 + i * 8.0;
    const tempValue = 28.0 + i * 0.9;

    const packets = [
      { location_id: locationId, type: 'Ultrasonic', value: ultrasonicValue },
      { location_id: locationId, type: 'Rain Intensity', value: rainValue },
      { location_id: locationId, type: 'Temperature', value: tempValue },
    ];

    console.log(`Push ${i}/${PUSH_COUNT} at t=${((Date.now() - start) / 1000).toFixed(2)}s`);
    for (const packet of packets) {
      const { status, body } = await postJson(endpoint, packet);
      const ok = status >= 200 && status < 300;
      console.log(`  ${ok ? 'OK' : 'ERR'} ${status} -> ${JSON.stringify(packet)}`);
      console.log(`    response: ${body}`);
    }
    console.log('');
  }

  const total = (Date.now() - start) / 1000;
  console.log(`Finished in ${total.toFixed(2)}s`);
}

void main();

